### PR TITLE
Example using 3D multichannel image.

### PR DESCRIPTION
**Pull request by mkcor**
## Description

I just started a gallery example where I'd like to show possible ways of interacting with 3D multichannel images. I'm using the kidney tissue fluorescence data by @GenevieveBuckley of course. :revolving_hearts: 

There's nothing interesting yet. The script does not even run as is, since #4939 isn't merged in yet.

I thought I would dive straight into the 3D challenges, then I realized that rendering the colours wasn't trivial: The image is closer to [BGR](https://scikit-image.org/docs/dev/user_guide/data_types.html#converting-bgr-to-rgb-or-vice-versa) than it is to RGB, but it's not exactly BGR either... :upside_down_face: 

## Checklist

- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- Gallery example in `./doc/examples` (new features only)
- Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- Unit tests
- Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

## For reviewers

- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

### CODE REVIEW

**Comment by GenevieveBuckley** [comment]
Thanks @mkcor!

**Comment by mkcor** [comment]
Now you understand why it's called confocal *fluorescence* microscopy!! w00t @GenevieveBuckley :heart: 
newplot
Achieved with:
```py
import plotly.express as px

px.imshow(data[n_plane // 2], zmin=v_min, zmax=v_max)
```
@emmanuelle @nicolaskruchten thanks for making ``imshow`` so powerful and flexible in Plotly Express!! I was looking at PR https://github.com/plotly/plotly.py/pull/2691 and I simply used https://plotly.com/python/imshow/#defining-the-data-range-covered-by-the-color-range-with-zmin-and-zmax (great documentation, btw).

I cannot achieve it with:
```py
from skimage import io

io.imshow(data[n_plane // 2], plugin='matplotlib', vmin=v_min, vmax=v_max)
```
fluo_fail
which roughly amounts to:
```py
import matplotlib.pyplot as plt

plt.imshow(data[n_plane // 2], vmin=v_min, vmax=v_max)
```
It's as if I didn't pass any min/max values...

**Comment by emmanuelle** [comment]
@mkcor yeah matplotlib's imshow uses `vmin/vmax` only for scalar (single-channel) data, it is ignored for multichannel data. We might checks with specialists (@GenevieveBuckley) if it's ok to change the contrast for each channel but since channels are not blended too much intermediate hues are not important. If it's confirmed that it's ok to change the contrast for fluorescence data then I'm really happy that we included this possibility in plotly's imshow. Maybe it's worth opening an issue in matplotlib's tracker to discuss this point?

**Comment by mkcor** [comment]
> We might check with specialists (@GenevieveBuckley) if it's ok to change the contrast for each channel but since channels are not blended too much intermediate hues are not important. If it's confirmed that it's ok to change the contrast for fluorescence data then I'm really happy that we included this possibility in plotly's imshow.

@emmanuelle yes, I would also prefer double-checking with a specialist of the field *if* and *how* we should change the contrast in a fluorescence microscopy image. Notably, should we do it on a *per channel* basis? I understand that my doing
```py
px.imshow(slice, zmin=v_min, zmax=v_max)
```
is equivalent to
```py
px.imshow(
    slice,
    zmin=[v_min, v_min, v_min],
    zmax=[v_max, v_max, v_max]
)
```
and, while the maximum value happens to be the same across channels (in this case, `4095`), the minimum value is not (it goes like `[10, 68, 35]`). Going for
```py
px.imshow(
    slice,
    zmin=[data[:, :, :, 0].min(), data[:, :, :, 1].min(), data[:, :, :, 2].min()],
    zmax=[v_max, v_max, v_max]
)
```
doesn't feel great, but maybe it's done...? I would say that
```py
px.imshow(
    slice,
    zmin=[0, 0, 0],
    zmax=[v_max, v_max, v_max]
)
```
looks most natural but, again, I'm not familiar enough with confocal fluorescence microscopy.

Let's poke some biology experts who might know better!
@GMFranceschini @slowkow @fran6w @ThomasWalter @kmader @IvoLeist

**Comment by mkcor** [comment]
> @mkcor yeah matplotlib's imshow uses `vmin/vmax` only for scalar (single-channel) data, it is ignored for multichannel data.

Oh, right, the [docs](https://matplotlib.org/api/_as_gen/matplotlib.pyplot.imshow.html) read "When using scalar data ..." indeed, but I didn't get that the parameter would just be ignored for multichannel data.

> Maybe it's worth opening an issue in matplotlib's tracker to discuss this point?

Ok, taking care of it now.

**Comment by GenevieveBuckley** [comment]
Changing the contrast independently for each channel for visualization is fine.

**Comment by kmader** [comment]
Great demo! I would also say it's fine to normalize on a per-channel basis and for most cases, it will prevent really poorly scaled images (why is my figure all black/white? questions) and having things just work out of the box is nice. 

I might have a note of it since many scikit-image tutorials end up as figures in papers (and comparing two groups with normalized channels might lead to wrong conclusions).

**Comment by mkcor** [comment]
@GenevieveBuckley @kmader thank you for your replies, they are really helpful!

> Great demo! I would also say it's fine to normalize on a per-channel basis and for most cases, it will prevent really poorly scaled images (why is my figure all black/white? questions) and having things just work out of the box is nice.

@kmader ok, in this case I will favour the following option:

```py
px.imshow(
    slice,
    zmin=[data[:, :, :, 0].min(), data[:, :, :, 1].min(), data[:, :, :, 2].min()],
    zmax=[data[:, :, :, 0].max(), data[:, :, :, 1].max(), data[:, :, :, 2].max()]
)
```

> I might have a note of it since many scikit-image tutorials end up as figures in papers (and comparing two groups with normalized channels might lead to wrong conclusions).

Oh, I would be very interested in this note! Thank you so much, again.

**Comment by fran6w** [comment]
@mkcor Thank you for the poke! Python, bioinformatics... but I am not a biology expert ;-)

**Comment by GMFranceschini** [comment]
@mkcor glad you thought about me! Unfortunately, I have never worked with images so I can't help with this issue!

**Comment by mkcor** [comment]
@fran6w @GMFranceschini I know your expertise is in bioinformatics rather than cell imaging... I reached out _just in case_ and we got answers anyway, so it's all good! Thanks anyway.

**Comment by mkcor** [comment]
Dear @GenevieveBuckley,

I'm wondering if each channel would correspond to specific (identifiable) tissues parts?
imshow_facet_col

PS: Happy New Year!! :tada:

**Comment by GenevieveBuckley** [comment]
> Dear @GenevieveBuckley,
> 
> I'm wondering if each channel would correspond to specific (identifiable) tissue parts?
> imshow_facet_col
> 
> PS: Happy New Year!! tada

Hi @mkcor - I must have missed this while I was off work over the new year. 

I don't think you can strictly say that certain fluorescence channels correspond with only certain specific tissue parts. One reason kidney tissue makes a good microscope test slide is that it has a huge amount of auto-fluorescence, so basically everything is going to look kinda bright regardless of your microscope settings. 

That said, the two bright blobs you can see best in channel two are very obviously kidney glomeruli
https://en.wikipedia.org/wiki/Glomerulus_(kidney)

I hope that helps!

**Comment by mkcor** [comment]
@GenevieveBuckley absolutely! It does help a lot. I have included your comments in the example (1671e9f).

@scikit-image/core please review. I'm happy to add an entry to the release notes, but I can't figure out which file under `doc/release/` I'm supposed to edit... :thinking:

**Comment by mkcor** [comment]
Thank you @JulesScholler for interacting on the [forum](https://forum.image.sc)! @GenevieveBuckley is a biologist and imaged the kidney tissue in question -- she answered my question by sharing the following:

> I don't think you can strictly say that certain fluorescence channels correspond with only certain specific tissue parts. One reason kidney tissue makes a good microscope test slide is that it has a huge amount of auto-fluorescence, so basically everything is going to look kinda bright regardless of your microscope settings.
> 
> That said, the two bright blobs you can see best in channel two are very obviously kidney glomeruli
> https://en.wikipedia.org/wiki/Glomerulus_(kidney)

**Comment by VolkerH** [comment]
> I don't think you can strictly say that certain fluorescence channels correspond with only certain specific tissue parts. One reason kidney tissue makes a good microscope test slide is that it has a huge amount of auto-fluorescence, so basically everything is going to look kinda bright regardless of your microscope settings.
> 

@GenevieveBuckley , but this is not just autofluorescence? I would assume that this was either immunofluorescence-labelled, the mouse was expressing fluorescent proteins or the tissue had some other fluorescent dye applied? Unfortunately that is not clear from the description of the dataset.

**Comment by GenevieveBuckley** [comment]
> @GenevieveBuckley is a biologist

Just to clarify I am NOT a biologist, and you should take anything I say about biology with a grain of salt. 

But I am confident that what I have said here is accurate. I've worked with groups imaging kidney tissue before, which is why I know what a glomerulus looks like. (Originally I studied physics, then worked as a scientist/computer programmer with research groups doing bio-medical or biological imaging. So my biology knowledge is pretty limited & only related to what those specific groups worked on.)

**Comment by GenevieveBuckley** [comment]
> > I don't think you can strictly say that certain fluorescence channels correspond with only certain specific tissue parts. One reason kidney tissue makes a good microscope test slide is that it has a huge amount of auto-fluorescence, so basically everything is going to look kinda bright regardless of your microscope settings.
> 
> @GenevieveBuckley , but this is not just autofluorescence? I would assume that this was either immunofluorescence-labelled, the mouse was expressing fluorescent proteins or the tissue had some other fluorescent dye applied? Unfortunately that is not clear from the description of the dataset.

It's not clear because I really don't know. The kidney slide & the lily of the valley slide were just test slides lying around when Judy showed me how to use one of the Nikkon inverted microscopes at MMI. Apparently the microscope companies often provide them, it certainly wasn't something made by a research group. We didn't care much about the specific sample, this was more a chance for me to learn how to use that particular microscope. It was only later that scikit-image were looking for more example images and I realised this was something I owned and didn't need anyone else's permission to be used.

You probably have a bit more experience in this area, what's your best guess about the sample slide?

**Comment by VolkerH** [comment]
Ah, ok . I thought it was a sample from the group who were counting glomeruli.
> 
> You probably have a bit more experience in this area, what's your best guess about the sample slide?

I really don't know. I would guess that the blue channel might be a nuclear marker such as Hoechst or DAPI.
But if it is not known with certainty I wouldn't want to add any additional information to the image, so there is no point guessing.

File: doc/examples/applications/plot_3d_interaction.py
Comment (GenevieveBuckley) [inline doc/examples/applications/plot_3d_interaction.py:10]:
```suggestion
The data we use here correspond to kidney tissue which was
```

File: doc/examples/applications/plot_3d_interaction.py
Comment (GenevieveBuckley) [inline doc/examples/applications/plot_3d_interaction.py:70]:
```suggestion
# Here you go, *fluorescence* microscopy!
```

File: doc/examples/applications/plot_3d_interaction.py
Comment (GenevieveBuckley) [inline doc/examples/applications/plot_3d_interaction.py:71]:
```suggestion

```

File: doc/examples/applications/plot_3d_interaction.py
Comment (GenevieveBuckley) [inline doc/examples/applications/plot_3d_interaction.py:72]:
```suggestion

```

File: doc/examples/applications/plot_3d_interaction.py
Comment (GenevieveBuckley) [inline doc/examples/applications/plot_3d_interaction.py:73]:
```suggestion

```

File: doc/examples/applications/plot_3d_interaction.py
Comment (GenevieveBuckley) [inline doc/examples/applications/plot_3d_interaction.py:144]: This bit is ok to leave in @mkcor

**Comment by mkcor** [comment]
> Just to clarify I am NOT a biologist, and you should take anything I say about biology with a grain of salt.

@GenevieveBuckley pardon my extrapolating! Sure, everything should always be taken with a grain of salt :wink: 

Thanks for the review! Note that you can now do [multi-line suggestions](https://github.blog/changelog/2020-02-26-multi-line-code-suggestions-beta/).

File: doc/examples/applications/plot_3d_interaction.py
Comment (jni) [inline doc/examples/applications/plot_3d_interaction.py:65]: Love this callout to the docs. =)

File: doc/examples/applications/plot_3d_interaction.py
Comment (jni) [inline doc/examples/applications/plot_3d_interaction.py:96]:
```suggestion
# What is the range of values for each colour channel?
# We check by taking the min and max across all non-channel
# axes.

vmin_0, vmin_1, vmin_2 = np.min(data, axis=(0, 1, 2))
vmax_0, vmax_1, vmax_2 = np.max(data, axis=(0, 1, 2))
```

File: doc/examples/applications/plot_3d_interaction.py
Comment (jni) [inline doc/examples/applications/plot_3d_interaction.py:112]:
```suggestion
# out, and exporting the desired figure as a static image in PNG format.
```

**Review by jni — APPROVED** [review]
@mkcor apologies for the delay in reviewing! This is a great little doc! :blush: I've made two very minor suggestions, but this is ok to merge with or without them. Thank you!

Since this is just a doc change I'll merge in ~24h unless anyone has objections.

File: doc/examples/applications/plot_3d_interaction.py
Comment (mkcor) [reply doc/examples/applications/plot_3d_interaction.py:96]: So elegant, love it! Thanks :)

**Comment by mkcor** [comment]
> I think just `doc/release/release_dev.rst`.

@grlee77 oh, thanks! 84d65a8da33bf04af5c I replaced the entry mentioning a previous biology example (#4648) with this current one, because #4648 made it into the [0.18 release](https://github.com/scikit-image/scikit-image/blob/main/doc/release/release_0.18.rst) so... I guess it should be pruned from `release_dev.rst`, once it lands in a given release? Not sure because a bunch of other Documentation entries can be found in both `release_dev.rst` and `release_0.18.rst`... :confused:

**Comment by jni** [comment]
@mkcor I think what's happened is that release_dev wasn't pruned after release. That would be my fault. Perhaps look at the log to see what happened in that file, if anything, since 0.18, and delete everything that was there at the 0.18 branch point. (https://github.com/scikit-image/scikit-image/commit/663e9e4b0f2ba88249bfef45bc65700555a3f24b, I think.)

**Comment by mkcor** [comment]
Hi @grlee77,

> My only question is about how we can make this and other Plotly-based examples more useful for users that try to run them from the command line.

Thank you for keeping this in mind! I admit I forgot about this.

> One possibility is calling `plotly.io.show` on each figure, for example.:
> 
> ```python
> fig = px.imshow(data[n_plane // 2], zmax=vmax)
> plotly.io.imshow(fig)
> ```

or simply (sparing `import plotly`):
```python
fig = px.imshow(data[n_plane // 2], zmax=vmax)
fig.show()
```

> If I do that everywhere in the example, I end up with one browser tab per plotly figure.

Right. I think it's quite nice, but...

> I didn't check if that will interfere with the gallery example, though? Let me know what you think.

It does! :sob: I tried and ended up with the same display issue as in: https://scikit-image.org/docs/dev/auto_examples/data/plot_3d.html It *used* to work, so I'm not sure when the issue arose (maybe a particular mix of Plotly and Sphinx versions...?). I'm noticing a display issue with the [thumbnail](https://scikit-image.org/docs/dev/auto_examples/index.html) as well.

> Adding a `plt.show()` at the end of the script can also keep the Matplotlib figure from disappearing immediately once the script completes. We have that in many of the existing Matplotlib-based examples.

Good point. In this gallery example, only one figure (i.e., the first one) is a Matplotlib figure... But it's better than nothing: done 1a0a7af3cf2b9a0f0c7691d996b042f64f8082f1.

**Comment by mkcor** [comment]
> @mkcor I think what's happened is that release_dev wasn't pruned after release. That would be my fault. Perhaps look at the log to see what happened in that file, if anything, since 0.18, and delete everything that was there at the 0.18 branch point. ([663e9e4](https://github.com/scikit-image/scikit-image/commit/663e9e4b0f2ba88249bfef45bc65700555a3f24b), I think.)

@jni ok, thanks. The overall cleanup could/should be a different PR, so I have pruned only the Documentation section (37b2b07aa3b45b, a0b0de7278cfb) by comparing `release_dev.rst` and `release_0.18.rst` side by side. I'll take care of item "Documentation has been added to the contributing notes about how to submit a gallery example" separately (either deleting it or finding corresponding PRs).

PR should be good to go now! :shipit:

**Review by grlee77 — APPROVED** [review]

**Comment by grlee77** [comment]
I opened #5341 about pruning the `release_dev.rst` file so we don't forget

**Comment by grlee77** [comment]
> I tried and ended up with the same display issue as in: https://scikit-image.org/docs/dev/auto_examples/data/plot_3d.html 

Is there an issue for that? Do you want to open a PR removing the call to show for now so things show up in the gallery as expected?

**Comment by mkcor** [comment]
> > I tried and ended up with the same display issue as in: https://scikit-image.org/docs/dev/auto_examples/data/plot_3d.html
> 
> Is there an issue for that? Do you want to open a PR removing the call to show for now so things show up in the gallery as expected?

Yes, I'll take care of it tomorrow (there's currently no issue for that).

**Pull request merged**
